examples/lighting-app/esp32/main/Button.cpp

```diff
@@ -15,52 +16,102 @@
+Button::Button(gpio_num_t gpioNum)
 {
-    if (button_press_handler != nullptr)
+    mGPIONum = gpioNum;
+}
+
+int32_t Find_Button_Via_Pin(gpio_num_t gpioNum)
+{
+    for (int i = 0; i < BUTTON_NUMBER; i++)
     {
-        button_press_handler();
+        if (gButtons[i].GetGPIONum() == gpioNum)
+        {
+            return i;
+        }
     }
+    return -1;
 }
 
-void Button::Init()
+void IRAM_ATTR button_isr_handler(void * arg)
 {
-    /* Initialize button interrupt*/
-    // zero-initialize the config structure.
+    uint32_t gpio_num = (uint32_t) arg;
+    int32_t idx       = Find_Button_Via_Pin((gpio_num_t) gpio_num);
+    if (idx == -1)
+    {
+        return;
+    }
+    BaseType_t taskWoken = pdFALSE;
+    xTimerStartFromISR(gButtons[idx].mbuttonTimer,
+                       &taskWoken); // If the timer had already been started ,restart it will reset its expiry time
+}
+
+esp_err_t Button::Init()
+{
+    return Init(mGPIONum);
+}
+
+esp_err_t Button::Init(gpio_num_t gpioNum)
+{
+    esp_err_t ret = ESP_OK;
+
+    mGPIONum = gpioNum;
+    //  zero-initialize the config structure.
     gpio_config_t io_conf = {};
-    // interrupt of rising edge
+    // interrupt of falling edge
     io_conf.intr_type = GPIO_INTR_NEGEDGE;
     // bit mask of the pins, use GPIO4/5 here
-    io_conf.pin_bit_mask = GPIO_INPUT_PIN_SEL;
+    io_conf.pin_bit_mask = 1ULL << gpioNum;
     // set as input mode
     io_conf.mode = GPIO_MODE_INPUT;
     // enable pull-up mode
     io_conf.pull_up_en = GPIO_PULLUP_ENABLE;
+
     gpio_config(&io_conf);
 
-    // install gpio isr service
-    gpio_install_isr_service(ESP_INTR_FLAG_DEFAULT);
     // hook isr handler for specific gpio pin
-    gpio_isr_handler_add(static_cast<gpio_num_t>(GPIO_INPUT_IO_0), gpio_isr_handler, (void *) GPIO_INPUT_IO_0);
+    ret = gpio_isr_handler_add(gpioNum, button_isr_handler, (void *) gpioNum);
+    ESP_RETURN_ON_ERROR(ret, TAG, "gpio_isr_handler_add failed: %s", esp_err_to_name(ret));
 
-    ESP_LOGI(TAG, "Button initialized..");
-}
+    mbuttonTimer = xTimerCreate("BtnTmr",               // Just a text name, not used by the RTOS kernel
+                                pdMS_TO_TICKS(50),      // timer period
+                                false,                  // no timer reload (==one-shot)
+                                (void *) (int) gpioNum, // init timer id = gpioNum index
+                                TimerCallback           // timer callback handler (all buttons use
+                                                        // the same timer cn function)
+    );
 
-void Button::SetButtonPressCallback(ButtonPressCallback button_callback)
+    return ESP_OK;
+}
+void Button::TimerCallback(TimerHandle_t xTimer)
 {
-    if (button_callback != nullptr)
-    {
-        button_press_handler = button_callback;
-    }
+    // Get the button index of the expired timer and call button event Handler.
+    uint32_t gpio_num = (uint32_t) pvTimerGetTimerID(xTimer);
```

Comment:
Do GPIOs of the M5 stack actually exceed 255? i don't think they do

```suggestion
    uint8_t gpio_num = (uint8_t) pvTimerGetTimerID(xTimer);
```

Reply:
probably not ... however docs for pvTimerGetTimerId seem to return void* ... I actually only copied this code from existing places so did not review/write it ... it probably is awkward.

Reply:
I believe we should have a cleanup phase once we merge button.cpp (and other) implementations. Not diverging them will make things easier to merge around (even though some things already obviously diverged ... like different LED widget with different capabilities)